src/x86_64/interrupt.c

```diff
@@ -323,6 +329,41 @@ void unregister_interrupt(int vector)
     interrupt_names[vector] = 0;
 }
 
+closure_function(1, 0, void, shirq_handler,
+                 list, handlers)
+{
+    list_foreach(bound(handlers), l) {
+        inthandler h = struct_from_list(l, inthandler, l);
+        int_debug("   invoking handler %s (%F)\n", h->name, h->t);
+        apply(h->t);
+    }
+}
+
+u64 allocate_shirq(void)
+{
+    u64 v = allocate_interrupt();
+    heap h = interrupt_vector_heap->meta;
+    list handlers = allocate(h, sizeof(struct list));
+    assert(handlers != INVALID_ADDRESS);
+    list_init(handlers);
+    thunk t = closure(h, shirq_handler, handlers);
+    assert(t != INVALID_ADDRESS);
+    register_interrupt(v, t, "shirq");
+    return v;
+}
+
+void register_shirq(int v, thunk t, const char *name)
```

Comment:
This isn't critical, but is there any particular reason to have another interface for shared ints rather than just allow multiple handler registrations per vector as aarch64 does? I guess the latter adds a dereference of the list head on dispatch even for non-shared ints, whereas this involves calling another handler to in turn dispatch shared handlers.

Reply:
I wrote the allocate_shirq() and register_shirq() functions before seeing the aarch64 implementation (but later I borrowed the inthandler struct from there), so I thought I would leave it as it is and see if it makes sense to you guys. Yes, the only advantage I see is that it avoids dereferencing the list head for non-shared interrupts. For performance-critical code we should try to avoid shared interrupts at all (if the VM allows some control over interrupt routing); I think the cost of invoking multiple handlers per interrupt and have them check the device status needlessly is much higher than the cost of having the "shirq" handler that dispatches the other handlers.